### PR TITLE
feat: support transaction

### PR DESCRIPTION
**Pull request by VaneLord67**
Fix: https://github.com/casbin/gorm-adapter/issues/175

support transaction
Example: 
```golang
  a := initAdapter(t, "mysql", "root:@tcp(127.0.0.1:3306)/", "casbin", "casbin_rule")
  e, _ := casbin.NewEnforcer("examples/rbac_model.conf", a)
  e.GetAdapter().(*Adapter).Transaction(e, func(e *casbin.Enforcer) error {
    _, err := e.AddPolicy("jack", "data1", "write")
    if err != nil {
	    return err
    }
    _, err = e.AddPolicy("jack", "data2", "write")
    //err = errors.New("some error")
    if err != nil {
	    return err
    }
    return nil
  })
```

### CODE REVIEW

**Comment by hsluoyz** [comment]
@VaneLord67 plz fix:

image

**Comment by hsluoyz** [comment]
@tangyang9464 @JalinWang plz review

**Comment by JalinWang** [comment]
Sorry for forgetting to submit and leaving the review pending for so long.
image

However, a key question is, `Enforcer` itself doesn't support transactions. We have no way to roll back in-memory policies.

**Comment by VaneLord67** [comment]
@JalinWang    the latest commit changes Enforcer to IEnforcer to support SyncedEnforcer，CachedEnforcer...； and reload policies from database at the end of transaction to refresh in-memory policies

**Comment by VaneLord67** [comment]
@JalinWang    the latest commit changes Enforcer to IEnforcer to support SyncedEnforcer，CachedEnforcer...； and reload policies from database at the end of transaction to refresh in-memory policies

**Comment by hsluoyz** [comment]
@tangyang9464

File: adapter.go [hunk 1]
```diff
@@ -611,6 +613,33 @@ func (a *Adapter) AddPolicies(sec string, ptype string, rules [][]string) error
 	return a.db.Create(&lines).Error
 }
 
+// Transaction perform a set of operations within a transaction
+func (a *Adapter) Transaction(e casbin.IEnforcer, fc func(casbin.IEnforcer) error, opts ...*sql.TxOptions) error {
```
Comment (hsluoyz) [inline adapter.go:617]: Plz also add to README about the new transaction usage

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Review by hsluoyz — APPROVED** [review]

**Comment by pinple** [comment]
how to use transaction between my business table and casbin_rule?

**Comment by hsluoyz** [comment]
@pinple see: https://github.com/casbin/gorm-adapter#transaction

**Comment by VaneLord67** [comment]
> 

```go
        a := initAdapter(t, "mysql", "root:@tcp(127.0.0.1:3306)/", "casbin", "casbin_rule")
	e, _ := casbin.NewEnforcer("examples/rbac_model.conf", a)
	mysqlDB := gorm.DB{}
	tx := mysqlDB.Begin()
	tx.Find(&struct {}{})
	err := e.GetAdapter().(*Adapter).Transaction(e, func(e casbin.IEnforcer) error {
		_, err := e.AddPolicy("jack", "data1", "write")
		if err != nil {
			return err
		}
		_, err = e.AddPolicy("jack", "data2", "write")
		if err != nil {
			return err
		}
		return nil
	})
	if err != nil {
		tx.Rollback()
	} else {
		tx.Commit()
	}
```

File: adapter.go [hunk 2]
```diff
@@ -611,6 +613,33 @@ func (a *Adapter) AddPolicies(sec string, ptype string, rules [][]string) error
...
+		err = e.LoadPolicy()
+		if err != nil {
+			panic(err)
+		}
+	}()
+	copyDB := *a.db
+	tx := copyDB.Begin(opts...)
+	b := &Adapter{db: tx}
+	// copy enforcer to set the new adapter with transaction tx
+	copyEnforcer := e
```
Comment (hulb) [inline adapter.go:632]: I'm curious here that this `e` seems to be a pointer and Won't any changes to `copyEnforcer` be applied to the origin `e`?

Comment (VaneLord67) [reply adapter.go:632]: Yes, you're correct. Therefore, I reset e's adapter in the defer function. The use of the copyEnforcer just makes the code more readable.

Comment (hulb) [reply adapter.go:632]: Got it.Thanks for reply!

**Pull request merged**